docs/versioning.md

```diff
@@ -0,0 +1,567 @@
+    longer the master. This option of versionId set to '' is used for creating
+    null versions once versioning has been suspended, which is discussed
+    further in "Null Version Management."
+
+Only one option is used at a time. `versionId: <versionId>` does not have to be
+used with `versioning: true` set to work, nor should they both be set. If both
+are used at once, the metadata engine will return an error.
+
+To summarize the valid combinations of versioning options:
+
+- !versioning && !versionId: normal non-versioning put
+- versioning && !versionId: create a new version, update the master version
+- !versioning && versionId: update (PUT/DELETE) an existing version
+  - if versionId === '' update master version
+
+Other cases are invalid and the metadata engine returns the error `BadRequest`.
+
+#### DELETE
+
+- no options: original DELETE operation, will delete the master version
+- `versionId: <versionId>` delete a specific version
+
+Deleting a specific version in metadata involves a few steps. A deletion
+targeting the latest version of an object has to:
+
+- delete the specified version identified by `versionId`
+- replace the master version with a version that is a placeholder for deletion
+  - this version contains a special keyword, 'isPHD', to indicate the master
+  version was deleted and needs to be updated
+- initiate a repair operation to update the value of the master version:
+  - involves listing the versions of the object and get the latest
+  version to replace the place holder delete version
+  - if no more versions exists, metadata deletes the master version, removing
+    the key from metadata
+
+Note: all of this happens before responding to S3, and only when the metadata
+engine is instructed by S3 to delete a specific version or the master version.
+See section "Delete Markers" for a description of what happens when a Delete
+Object request is sent to the S3 API.
+
+#### GET
+
+- no options: original GET operation, will get the master version
+- `versionId: <versionId>` retrieve a specific version
+
+The implementation of a GET operation does not change compared to the standard
+version. A standard GET without versioning information would get the master
+version of a key. A version-specific GET would retrieve the specific version
+identified by the key for that version.
+
+#### LIST
+
+For a standard LIST on a bucket, metadata iterates through the keys by using
+the separator represented by `.` as an extra delimiter. For a listing of all
+versions of a bucket, there is no change compared to the original listing
+function. Instead, the API component returns all the keys in a List Objects call
+and filters for just the keys of the master versions in a List Object Versions
+call.
+
+For example, a standard LIST operation against the keys in a table below would
+return from metadata the list of `[ foo/bar, bar, qux/quz, quz ]` using the
+separator represented by `.` as the extra delimiter.
```

Comment:
Yes, I think it's a little clearer to read when using the `.` in the examples than putting the actual `\0` as long as it's understood the `.` is representative of the actual separator (`\0`).